gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1148 @@
+        DISABLED,               // dividend_amt
+        DISABLED,               // capg_amt
+        "Reverse split",
+        ""
+    },
+    {
+        ENABLED_DEBIT,          // stock_amt
+        ENABLED_DEBIT,          // stock_val
+        ENABLED_CREDIT,         // cash_amt
+        ENABLED_DEBIT | ALLOW_ZERO,          // fees_amt
+        false,                  // fees_capitalize
+        DISABLED,               // dividend_amt
+        ENABLED_DEBIT | ALLOW_ZERO | ALLOW_NEGATIVE,          // capg_amt
+        "Reverse split w/ cash in lieu for fractionals",
+        ""
+    },
+};
+
+typedef struct
+{
+    GtkWidget * window;
+    GtkWidget * assistant;
+
+    TxnTypeVec txn_types;
+    Account   * acct;
+    gnc_commodity * currency;
+
+    // transaction type page
+    GtkWidget * transaction_type_page;
+    GtkWidget * transaction_type_combo;
+    GtkWidget * transaction_type_explanation;
+    TxnTypeInfo txn_type;
+
+    // transaction details page
+    GtkWidget * transaction_details_page;
+    GtkWidget * date_edit;
+    GtkWidget * transaction_description_entry;
+
+    // stock amount page
+    gnc_numeric balance_at_date;
+    GtkWidget * stock_amount_page;
+    GtkWidget * prev_amount;
+    GtkWidget * next_amount;
+    GtkWidget * stock_amount_edit;
+
+    // stock value page
+    GtkWidget * stock_value_page;
+    GtkWidget * stock_value_edit;
+    GtkWidget * price_value;
+    GtkWidget * stock_memo_edit;
+
+    // cash page
+    GtkWidget * cash_page;
+    GtkWidget * cash_account;
+    GtkWidget * cash_memo_edit;
+    GtkWidget * cash_value;
+
+    // fees page
+    GtkWidget * fees_page;
+    GtkWidget * capitalize_fees_checkbox;
+    GtkWidget * fees_account;
+    GtkWidget * fees_memo_edit;
+    GtkWidget * fees_value;
+
+    // dividend page
+    GtkWidget * dividend_page;
+    GtkWidget * dividend_account;
+    GtkWidget * dividend_memo_edit;
+    GtkWidget * dividend_value;
+
+    // capgains page
+    GtkWidget * capgains_page;
+    GtkWidget * capgains_account;
+    GtkWidget * capgains_memo_edit;
+    GtkWidget * capgains_value;
+
+    // finish page
+    GtkWidget * finish_page;
+    GtkWidget * finish_split_view;
+    GtkWidget * finish_summary;
+} StockTransactionInfo;
+
+
+/******* implementations ***********************************************/
+static void
+stock_assistant_window_destroy_cb (GtkWidget *object, gpointer user_data)
+{
+    StockTransactionInfo *info = (StockTransactionInfo *)user_data;
+    gnc_unregister_gui_component_by_data (ASSISTANT_STOCK_TRANSACTION_CM_CLASS, info);
+    g_free (info);
+}
+
+static void refresh_page_transaction_type (GtkWidget *widget, gpointer user_data)
+{
+    StockTransactionInfo *info = (StockTransactionInfo *)user_data;
+
+    auto type_idx = gtk_combo_box_get_active (GTK_COMBO_BOX (widget));
+    if (type_idx < 0)           // combo isn't initialized yet.
+        return;
+
+    info->txn_type = info->txn_types.at (type_idx);
```

Comment:
This currently leaks:
```
==38937== 14 bytes in 1 blocks are definitely lost in loss record 1,173 of 19,428
==38937==    at 0x4843FB3: operator new(unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==38937==    by 0x62171CE: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29)
==38937==    by 0x621752D: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator=(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29)
==38937==    by 0x4D96B62: TxnTypeInfo::operator=(TxnTypeInfo const&) (assistant-stock-transaction.cpp:95)
==38937==    by 0x4D91B36: refresh_page_transaction_type(_GtkWidget*, void*) (assistant-stock-transaction.cpp:436)
==38937==    by 0x5AAA96F: g_signal_emit_valist (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937==    by 0x5AAAAD2: g_signal_emit (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937==    by 0x51F0C06: ??? (in /usr/lib/x86_64-linux-gnu/libgtk-3.so.0.2404.26)
==38937==    by 0x51F0DD2: gtk_combo_box_set_active (in /usr/lib/x86_64-linux-gnu/libgtk-3.so.0.2404.26)
==38937==    by 0x4D9332C: stock_assistant_prepare (assistant-stock-transaction.cpp:701)
==38937==    by 0x5A8B624: g_cclosure_marshal_VOID__OBJECTv (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937==    by 0x5AAA96F: g_signal_emit_valist (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
```

Reply:
I *tried* redefine `txn_type` and `txn_types` as

```cpp
TxnTypeInfo * txn_type;
TxnTypeVec * txn_types;
```

but doesn't compile. This _may_ require some fancy `shared_ptr`?